### ISSUE TITLE
Nextgen Core: Caching

### ISSUE THREAD

**Issue by Makman2** [coala/coala#4350]
The new core has currently no support for caching. We need to think of some neat mechanism to provide caching.

**Comment by Makman2** [comment]
My current idea:
- Use the parameters received from `generate_tasks()` and cache those with the results. We need to check how disk-space usage can be optimized, as the tasks do contain all files usually and its contents. Also the current `ProjectBear` and `FileBear` implementations do optimize performance a bit at `generate_tasks()` to not provide all arguments with `generate_tasks()` but grab them from the class instance instead. Needs investigation whether this can be kept (for example by always saving the associated section with each "task-object" from `generate_tasks()`) or not.

**Comment by Makman2** [comment]
I wrapped my head now a week literally in total to think everything through (so 7 * 24 hours), especially in context with hashing.

To clarify again: What is the duty of the cache? The cache shall check if we have executed a bear already, more precisely a single task. In case the task has been executed before, cached `Result`s are retrieved immediately, and `analyze` won't be run again.

We can assume that coala will handle for a single run a few tasks minimum (at least 1), and in the "worst-case" about 1 million (for really large projects, assumed 10000 files and a huge setup with 100 `FileBear`s). Like mentioned, the idea was to directly store the task objects (the arguments to invoke `analyze` with) inside a dict, that functions as a cache. It maps the task objects to the already computed results from a previous run.

There is one big problem with this approach: Currently for `FileBear`s, we pass each file to analyze inside the task. This effectively means, we store all files again we want to analyze with coala in the cache. This produces a cache taking a huge amount of disk space and RAM. Even if a user decides that he has enough disk space and RAM (which is nowadays very probable), coala will anyway get a huge performance penalty. Loading the cache from a file (that has to be done to make the cache persistent across coala runs) takes time.
Also in our current design, `ProjectBear`s do pass a single task to a coala run, and it contains a dictionary (the file-dict). dicts don't work as keys inside our cache-dict, so without some tweaks we can't do this directly.

This problem might be solved by introducing the planned file-factory: `FileBear`s are passed a file-factory object instead of the file-contents and -name directly. We can modify the class to be treated equal amongst other instances of the same class, if only the *timestamp* of the file referenced is the same. This will reduce the cache size insanely, as we don't have to store the actual file-contents, but just the timestamps (and file names).

However, this leads to the next step: How can the cache be made smart, so it can handle such "object-identifying" values? This is pretty related to hashing. If a hash of two objects are equal, they can be seen as equal objects (hash collisions aside, coming to that later). Citing Python docs:

> The only required property is that objects which compare equal have the same hash value; it is advised to mix together the hash values of the components of the object that also play a part in comparison of objects by packing them into a tuple and hashing the tuple.

So these members which are mixed are the identifying values of an object that defines it completely (ideally). So we could store the hashes of each object in the cache and compare them later with those in later coala runs. In fact we could just hash the task itself. But there are devastating problems:

1. We pass unhashable objects around (see above with `ProjectBear` and file-dict)
2. From Python docs:
   > By default, the __hash__() values of str, bytes and datetime objects are “salted” with an unpredictable random value. Although they remain constant within an individual Python process, they are not predictable between repeated invocations of Python.
   >
   >
This is intended to provide protection against a denial-of-service caused by carefully-chosen inputs that exploit the worst case performance of a dict insertion, O(n^2) complexity. See http://www.ocert.org/advisories/ocert-2011-003.html for details.

So hashes aren't persistent for `str`, `bytes` and `datetime`. So hashing the task objects doesn't work either without tweaks.

I thought about following next possibilities:
- Implement tweaks for those cases, defining an own "hash"-method.
  This is quite problematic with unhashable values still.
- Using `hashlib` for reliable and persistent hashing.

Because unhashable values are still a huge problem with method 1, I researched further on method 2. Specifically I looked at sha1, which seems to be quite fast. One major drawback: The hashlib algorithms can generate the hash with `bytes` input. So we would have to serialize each object before we can generate a hash here. I somehow stumbled across the idea to use `pickle.dumps`.

Thinking over this again, this seemed to me the perfect approach: `pickle` serializes data that identifies the object, like hash values do. Even unhashable items do work (conside a dict: The contents of a dict define the dict, it's not hashable because it's just mutable). And because the normal `hash` function in Python doesn't produce persistent hashes for `bytes` generated by `pickle.dumps`, we use an alternative hashing algorithm from hashlib (sha1). We now have the major advantage that we can now cache any object we like, and even quite simple and easy (knowing that all now).

There's only one drawback again (which doesn't make this solution explode though): Performance. I'm not sure how big the performance impact will be using coala with a cache. Now we have to:
1. Load the cache when starting coala (probably taking an inconsiderable amount of time, now that we only save "a few" 160-bit hash values)
2. Hash each task object that coala's core is fed with, that means:
   2.1. pickle the task object
   2.2. invoke the sha1 hash algorithm
   2.3. perform cache-lookup

It could be that for some cases caching has to be disabled manually sometimes by the developer, because the algorithm is so trivial it doesn't make sense to cache it (maybe this is the case for `SpaceConsistencyBear` / similar algorithms). But that is something that can and should be considered after an implementation with empirical results.

Worth to note: Using sha1 hashes makes coala's cache not working "ideal". That means it's possible that due to a very bad coincidence, coala returns wrong results even with an intact cache. As you know, hashing maps objects of arbitrary size to a fixed-size number. There's is no way that there can't be two different objects that have the same hash, i.e. hash-values can collide. Python's `hash` has two different bit-lengths for hash-values, depending on the platform: 32 and 64. For x86, we have a guaranteed collision with the 1 million tasks worst-case from above, with x64 platforms the collision probability is quite low (2.7e-8). sha1 bahaves better with a hash-value-length of 160bit: Even in the worst case we can assume that a collision is statistically impossible with 3.4e-37. (I left out other non-ideal details of hash-algorithms that probably raise all probabilities by some amount, e.g. non-linearity and input-type-probability. I assume a rise of 2-3 orders of magnitude.)
That means: sha1 is definitely the better choice than Python's `hash`.

**Comment by Makman2** [comment]
One problem I see just now: `pickle` might not preserve data-order for `set`, `dict` and so on, so cache entries might become invalid just because objects are stored slightly different in the next Python execution. For such cases we have to patch the pickling-functions to produce always the same data-order.

Some problems:
1. How can this be reliably tested (that data-order is reliably preserved)?
2. If someone uses self-written objects having the same behaviour as `set` and `dict` (producing different bytes for two objects considered to be equal), how can our bear-caching be made aware of this special object and influence pickling?

I propose to do 1. in a separate issue, and do some manual testing of caching first. That should suffice for the first implementation.

**Comment by Makman2** [comment]
I thought about abstracting away the calculation of our persistent cache hashes away into an own function. The function is responsible for generating hashes that our bear can reliably use. It's just a shame that depending on an object's nature, a user has maybe to modify the behaviour, Python can't do everything automatically here. Following behaviours shall the function respect:

1. By default: Use `pickle.dumps` and `sha1` as hash algorithm on the pickled data stream. This works on most objects.
2. For objects that are unhashable:
   - Implement `hash()`
   - For built-in types: Custom pickle functions have to be registered for those. For built-in types, they are hard-coded into the function.
   - For user types: We need a system to register pickling functions like `pickle` does or allow to override a special method that we invoke to retrieve the hash value or serialized data.
3. For objects whose hashes depend on the runtime (cache can hash them by default with method 1, but the cache entries will be always missed):
   - For built-in types: Like with method 2, hard-code the right pickling behaviour for those.
   - For user types: Again like with method 2.
